[Arch] Backintime fails to build

Hi, I have an up to date Arch install and I am trying install the git version of backintime via the AUR using yay. But the install fails during unittesting for reason. Here is the output when it fails

OK
python3 -m unittest  -b test/test_backintime.py
E..
======================================================================
ERROR: test_diagnostics_arg (test.test_backintime.TestBackInTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/test/test_backintime.py", line 181, in test_diagnostics_arg
    diagnostics = json.loads(output)
  File "/usr/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

----------------------------------------------------------------------
Ran 3 tests in 3.040s

Still fails to build for me on Arch. Though the error is different now

`python3 -m unittest  -b test/test_backintime.py
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 1190, in <module>
    startApp()
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 507, in startApp
    args = argParse(None)
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 568, in argParse
    args, unknownArgs = mainParser.parse_known_args(args)
  File "/usr/lib/python3.10/argparse.py", line 1859, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.10/argparse.py", line 2068, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/usr/lib/python3.10/argparse.py", line 2008, in consume_optional
    take_action(action, args, option_string)
  File "/usr/lib/python3.10/argparse.py", line 1936, in take_action
    action(self, namespace, argument_values, option_string)
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/backintime.py", line 742, in __call__
    diagnostics = collect_diagnostics()
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/diagnostics.py", line 80, in collect_diagnostics
    result['host-setup']['locale'] = ', '.join(locale.getlocale())
TypeError: sequence item 0: expected str instance, NoneType found
E.s
======================================================================
ERROR: test_diagnostics_arg (test.test_backintime.TestBackInTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/af/.cache/yay/backintime-git/src/backintime/common/test/test_backintime.py", line 191, in test_diagnostics_arg
    output = subprocess.check_output(["./backintime", "--diagnostics"])
  File "/usr/lib/python3.10/subprocess.py", line 421, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['./backintime', '--diagnostics']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 3 tests in 2.966s

FAILED (errors=1, skipped=1)
make: *** [Makefile:570: unittest] Error 1
make: Leaving directory '/home/af/.cache/yay/backintime-git/src/backintime/common'`

--- a/common/diagnostics.py
+++ b/common/diagnostics.py
@@ -138,7 +138,7 @@ def collect_diagnostics():
     # rsync <= (somewhere near) 3.1.3: -VV doesn't exists
     # rsync == 3.1.3 (Ubuntu 20 LTS) doesn't even know '-V'
 
-    # This work when rsync understand -VV and return json or human readable
+    # This works when rsync understands -VV and returns json or human readable
     result['external-programs']['rsync'] = _get_extern_versions(
         ['rsync', '-VV'],
         r'rsync  version (.*)  protocol version',
@@ -189,15 +189,19 @@ def _get_extern_versions(cmd,
                          error_pattern=None):
     """Get the version of an external tools using ``subprocess.Popen()``.
 
-    Args:
-        cmd (list): Commandline arguments that will be passed to `Popen()`.
-        pattern (str): A regex pattern to extract the version string from the
-                       commands output.
-        try_json (bool): Interpet the output as json first (default: False).
-        error_pattern (str): Regex pattern to identify a message in the output
-                             that indicates an error.
+     Args:
+         cmd (list): Commandline arguments that will be passed to `Popen()`.
+         pattern (str): A regex pattern to extract the version string from the
+                        commands output.
+         try_json (bool): Interpret the output as json first (default: False).
+                          If it could be parsed the result is a dict
+         error_pattern (str): Regex pattern to identify a message in the output
+                              that indicates an error.
 
-    """
+     Returns:
+         Version information as string (or dict if was JSON and parsed successfully).
+         `None` if the error_pattern did match (to indicate an error)
+     """
 
     try:
         # as context manager to prevent ResourceWarning's
@@ -219,7 +223,7 @@ def _get_extern_versions(cmd,
             if match:
                 return None
 
-        # some tools use "stderr" for version infos
+        # some tools use "stderr" for version info
         if not std_output:
             result = error_output
         else:
@@ -236,14 +240,13 @@ def _get_extern_versions(cmd,
                 pass
 
             else:
-                # No regex parsing because it was json
-                pattern = None
+                return result  # as JSON
 
         # extract version string
         if pattern:
             result = re.findall(pattern, result)[0]
 
-    return result.strip()
+    return result.strip()  # as string
 
 
 def get_git_repository_info(path=None):

--- a/common/diagnostics.py
+++ b/common/diagnostics.py
@@ -77,7 +77,14 @@ def collect_diagnostics():
         'XDG_SESSION_TYPE', '($XDG_SESSION_TYPE not set)')
 
     # locale (system language etc)
-    result['host-setup']['locale'] = ', '.join(locale.getlocale())
+    #
+    # Implementation note: With env var "LC_ALL=C" getlocale() will return (None, None).
+    # This throws an error in "join()":
+    #   TypeError: sequence item 0: expected str instance, NoneType found
+    my_locale = locale.getlocale()
+    if all(x is None for x in my_locale):
+        my_locale = ["(Unknown)"]
+    result['host-setup']['locale'] = ', '.join(my_locale)
 
     # PATH environment variable
     result['host-setup']['PATH'] = os.environ.get('PATH', '($PATH unknown)')

--- a/common/test/test_backintime.py
+++ b/common/test/test_backintime.py
@@ -129,8 +129,21 @@ This is free software, and you are welcome to redistribute it
 under certain conditions; type `backintime --license' for details.
 ''', re.MULTILINE))
 
-        # The log output completely goes to stderr
-        self.assertRegex(error.decode(), re.compile(r'''INFO: Lock
+        # The log output completely goes to stderr.
+        # Note: DBus warnings at the begin and end are already ignored by the regex
+        #       but if the BiT serviceHelper.py DBus daemon is not installed
+        #       at all the warnings also occur in the middle of below expected
+        #       INFO log lines so they are removed by filtering here.
+        # TODO If filtering output rows should become a common testing pattern
+        #      this code should be refactored into a function for reusability.
+        log_output = []
+        for line in error.decode().split("\n"):
+            if (not line.startswith("WARNING: Failed to connect to Udev serviceHelper")
+                    and not line.startswith("WARNING: D-Bus message:")
+                    and not line.startswith("WARNING: Udev-based profiles cannot be changed or checked")):
+                log_output.append(line)
+        filtered_log_output = "\n".join(log_output)
+        self.assertRegex(filtered_log_output, re.compile(r'''INFO: Lock
 INFO: Take a new snapshot. Profile: 1 Main profile
 INFO: Call rsync to take the snapshot
 INFO: Save config file

--- a/common/test/test_backintime.py
+++ b/common/test/test_backintime.py
@@ -140,7 +140,8 @@ under certain conditions; type `backintime --license' for details.
         for line in error.decode().split("\n"):
             if (not line.startswith("WARNING: Failed to connect to Udev serviceHelper")
                     and not line.startswith("WARNING: D-Bus message:")
-                    and not line.startswith("WARNING: Udev-based profiles cannot be changed or checked")):
+                    and not line.startswith("WARNING: Udev-based profiles cannot be changed or checked")
+                    and not line.startswith("WARNING: Inhibit Suspend failed")):
                 log_output.append(line)
         filtered_log_output = "\n".join(log_output)
         self.assertRegex(filtered_log_output, re.compile(r'''INFO: Lock

--- a/qt/qttools.py
+++ b/qt/qttools.py
@@ -179,6 +179,14 @@ def createQApplication(app_name = 'Back In Time'):
         qapp.style().objectName().lower() == 'windows' and  \
         'GTK+' in QStyleFactory.keys():
             qapp.setStyle('GTK+')
+
+    try:
+        # The platform name indicates eg. wayland vs. X11, see also:
+        # https://doc.qt.io/qt-5/qguiapplication.html#platformName-prop
+        logger.debug(f"QT QPA platform plugin: {qapp.platformName()}")
+    except Exception as e:
+        logger.debug(f"QT QPA platform plugin unknown: {repr(e)}")
+
     return qapp
 
 def translator():

